chassembly/src/main/java/org/quiltmc/chasm/lang/internal/parse/Parser.jj

```diff
@@ -627,13 +627,25 @@ TOKEN: {
     |
     <BooleanLiteral: "true" | "false">
     |
-    <DecIntegerLiteral: ("+" | "-")? (["0" - "9"])+ >
+    <#Sign: "+" | "-">
     |
-    <HexIntegerLiteral: "0x" (["0" - "9", "a" - "f", "A"-"F"])+ >
+    <#Digit: ["0" - "9"]>
     |
-    <BinIntegerLiteral: "0b" (["0" - "1"])+ >
+    <#HexDigit: <Digit> | ["a" - "f", "A" - "F"] >
     |
-    <FloatLiteral: ("+" | "-")? (["0" - "9"])+ "." (["0" - "9"])+ ("e" ("+" | "-")? (["0" - "9"])+)?>
+    <DecIntegerLiteral: (<Sign>)? (<Digit>)+ >
+    |
+    <HexIntegerLiteral: (<Sign>)? "0x" (<HexDigit>)+ >
+    |
+    <BinIntegerLiteral: (<Sign>)? "0b" (["0" - "1"])+ >
+    |
+    <#SpecialFloatLiteral: (<Sign>)? ("NaN" | "Infinity")>
+    |
+    <#DecFloatLiteral: (<Sign>)? (((<Digit>)+ "." (<Digit>)*) | ("." (<Digit>)+)) (("e" | "E") (<Sign>)? (<Digit>)+)? >
+    |
+    <#HexFloatLiteral: (<Sign>)? "0x" (((<HexDigit>)+ ("." (<HexDigit>)*)?) | ("." (<HexDigit>)+)) ("p" | "P") (<Sign>)? (<Digit>)+>
```

Comment:
I am not convinced by hex float literals.
I don't see any value in them being added.

Reply:
Some values are easier to express in hex, such as large powers of 2. Plus it's parity with Java literals so why not?

Reply:
I don't think "compatibility with java" is a good argument. I've also decided against octal integers.
I guess you could argue that `0x1p32` is convenient, but how about we just use `1<<32` instead?
In the rare case that you feel the need to do `0x1p-5`, i think most people would prefer if you just write the division.
I don't think this is good or even useful syntax, and I would prefer to not add bad syntax.

Reply:
> but how about we just use `1<<32` instead? 

Because that only makes sense for integers. You could argue that you could convert that to a float, but then what about `0x1p64`? That's out of range for integers.

Reply:
Discussion moved to #85